File: Data Structures & Algorithms/course-schedule/submission-2.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] course = new int[numCourses];
        List<List<Integer>> adj = new ArrayList<>();

        for (int i = 0; i < numCourses; i++) {
            adj.add(new ArrayList<>());
        }
        for (int[] pre : prerequisites) {
            course[pre[1]]++;
            adj.get(pre[0]).add(pre[1]);
        }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (course[i] == 0) {
                q.add(i);
            }
        }

        int finish = 0;

        while (!q.isEmpty()) {
            int node = q.poll();
            finish++;
            for (int nei : adj.get(node)) {
                course[nei]--;
                if (course[nei] == 0)
                    q.add(nei);
            }
        }

        return finish == numCourses;
    }
}

File: Data Structures & Algorithms/course-schedule-iv/submission-2.py
class Solution:
    def checkIfPrerequisite(
        self, numCourses: int, prerequisites: List[List[int]], queries: List[List[int]]
    ) -> List[bool]:
        adj = [[] for _ in range(numCourses)]
        isPre = [[-1] * numCourses for _ in range(numCourses)]

        for pre, crs in prerequisites:
            adj[crs].append(pre)
            isPre[crs][pre] = True

        def dfs(crs, pre):
            if isPre[crs][pre] != -1:
                return isPre[crs][pre] == 1

            for p in adj[crs]:
                if p == pre or dfs(p, pre):
                    isPre[crs][pre] = 1
                    return True

            isPre[crs][pre] = 0
            return False

        res = []
        for u, v in queries:
            res.append(dfs(v, u))

        return res
